src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,379 @@
+                now = System.currentTimeMillis();
+            }
+            if (lockedEventTypes.contains(eventType)) {
+                throw new TimeoutException("Timed out while waiting for event type " + eventType +
+                        " to unlock within " + timeoutMillis);
+            }
+            eventsBeingPublished.put(eventType, eventsBeingPublished.getOrDefault(eventType, 0) + 1);
+        }
+        return () -> {
+            synchronized (localLock) {
+                final int currentCount = eventsBeingPublished.get(eventType);
+                if (1 == currentCount) {
+                    eventsBeingPublished.remove(eventType);
+                    localLock.notifyAll();
+                } else {
+                    eventsBeingPublished.put(eventType, currentCount - 1);
+                }
+            }
+        };
+    }
+
+    private void updateVersionAndWaitForAllNodes(@Nullable final Long timeoutMs) throws InterruptedException {
+        // Create next version, that will contain locked event type.
+        final Optional<Long> expectedFinish = Optional.ofNullable(timeoutMs).map(t -> System.currentTimeMillis() + t);
+        final AtomicInteger versionToWait = new AtomicInteger();
+        runLocked(() -> {
+            try {
+                final int latestVersion = readData("/version", null, Integer::valueOf);
+                versionToWait.set(latestVersion + 1);
+                zooKeeperHolder.get().setData().forPath(
+                        toZkPath("/version"), String.valueOf(versionToWait.get()).getBytes(Charsets.UTF_8));
+                LOG.info("Wrote to ZK version to wait (was: {}), new one: {}", latestVersion, versionToWait.get());
+            } catch (final RuntimeException ex) {
+                throw ex;
+            } catch (final Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+        // Wait for all nodes to have latest version.
+        final AtomicBoolean latestVersionIsThere = new AtomicBoolean(false);
+        while (!latestVersionIsThere.get()) {
+            LOG.info("Waiting for all nodes to have the same version {}", versionToWait.get());
+            if (expectedFinish.isPresent() && System.currentTimeMillis() > expectedFinish.get()) {
+                throw new RuntimeException("Timed out while updating version to " + versionToWait.get());
+            }
+            Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+            runLocked(() -> {
+                try {
+                    final List<String> nodes = zooKeeperHolder.get().getChildren().forPath(toZkPath("/nodes"));
+                    boolean allOk = true;
+                    for (final String node : nodes) {
+                        final int nodeVersion = readData("/nodes/" + node, null, Integer::valueOf);
+                        final boolean nodeOk = nodeVersion >= versionToWait.get();
+                        if (!nodeOk) {
+                            allOk = false;
```

Comment:
@v-stepanov Initially (before tests) it was with break. But now it is like that because of additional logging (which makes sence while switching)